### PR TITLE
James 2617 ElasticSearch 6 metrics

### PR DESCRIPTION
**Pull request by rouazana**
Here is a PR to make James metrics compatible with Elastic Search 6.

I was able to manually test that it's working fine with ES 6 and ES 2.

For the moment I was not able to update the integration tests because they rely on ES 2 protocol, I need to find a way to make them compatible with ES 6 and ES 2 at the same time.

### CODE REVIEW

File: third-party/pom.xml [hunk 1]
```diff
@@ -35,6 +35,7 @@
 
     <modules>
         <module>spamassassin</module>
+        <module>nerdynick-elasticsearch-metrics-reporter</module>
```
Comment (mbaechler) [inline third-party/pom.xml]: alphabetical ordering ?

File: third-party/nerdynick-elasticsearch-metrics-reporter/pom.xml [hunk 2]
```diff
@@ -0,0 +1,255 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (chibenwa) [inline third-party/nerdynick-elasticsearch-metrics-reporter/pom.xml]: Missing license

File: third-party/nerdynick-elasticsearch-metrics-reporter/pom.xml [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    
+    <groupId>org.elasticsearch</groupId>
+    <artifactId>metrics-elasticsearch-reporter</artifactId>
+    <version>6.0.0-SNAPSHOT-apache-james</version>
```
Comment (chibenwa) [inline third-party/nerdynick-elasticsearch-metrics-reporter/pom.xml]: We should not introduce this kind of SNAPSHOT dependencies as it makes releasing harder.

File: third-party/nerdynick-elasticsearch-metrics-reporter/pom.xml [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+            <artifactId>jackson-databind</artifactId>
+            <version>${jackson.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.module</groupId>
+            <artifactId>jackson-module-afterburner</artifactId>
+            <version>${jackson.version}</version>
+        </dependency>
+        <dependency>
+			<groupId>org.apache.spark</groupId>
```
Comment (chibenwa) [inline third-party/nerdynick-elasticsearch-metrics-reporter/pom.xml]: Ident warning

Comment (chibenwa) [reply third-party/nerdynick-elasticsearch-metrics-reporter/pom.xml]: Can't we remove this spark dependency?

File: third-party/nerdynick-elasticsearch-metrics-reporter/pom.xml [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+        </dependency>
+        <dependency>
+            <groupId>com.carrotsearch.randomizedtesting</groupId>
+            <artifactId>randomizedtesting-runner</artifactId>
+            <version>${randomized.testrunner.version}</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-simple</artifactId>
```
Comment (chibenwa) [inline third-party/nerdynick-elasticsearch-metrics-reporter/pom.xml]: Use logback classic instead?

File: third-party/nerdynick-elasticsearch-metrics-reporter/samples/simple-webapp/src/main/java/web/notifier/HttpNotifier.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+import org.elasticsearch.metrics.percolation.Notifier;
+
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.net.URL;
+
+/**
+ * Send a post request to the cockpit sample application, so it gets shown as a notification in there
+ */
+public class HttpNotifier implements Notifier {
```
Comment (chibenwa) [inline third-party/nerdynick-elasticsearch-metrics-reporter/samples/simple-webapp/src/main/java/web/notifier/HttpNotifier.java]: Are these samples useful to add to James source code?

File: third-party/nerdynick-elasticsearch-metrics-reporter/pom.xml [hunk 7]
```diff
@@ -0,0 +1,255 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    
+    <groupId>org.elasticsearch</groupId>
+    <artifactId>metrics-elasticsearch-reporter</artifactId>
+    <version>6.0.0-SNAPSHOT-apache-james</version>
+
+    <properties>
+        <lucene.version>5.5.0</lucene.version>
```
Comment (chibenwa) [inline third-party/nerdynick-elasticsearch-metrics-reporter/pom.xml]: Isn't it a good opportunity to align depencies versions of this lib?

```
Jackson 2.7.3 -> 2.9.6
Metrics 3.1.2 -> 3.2.6
Guava 24.1 -> 25.1
```

File: third-party/nerdynick-elasticsearch-metrics-reporter/pom.xml (hunk 3 above)
Comment (chibenwa) [reply third-party/nerdynick-elasticsearch-metrics-reporter/pom.xml]: BTW I fear that we will not be able to release this. Can't we change the groupId to `org.apache.james` ?

File: third-party/nerdynick-elasticsearch-metrics-reporter/pom.xml [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+        <elasticsearch.version>2.3.1</elasticsearch.version>
+        <jackson.version>2.7.3</jackson.version>
+        <randomized.testrunner.version>2.3.3</randomized.testrunner.version>
+        <spark.version>2.2.0</spark.version>
+    </properties>
+
+    <modelVersion>4.0.0</modelVersion>
+    <packaging>jar</packaging>
+    <name>metrics-elasticsearch-reporter</name>
+    <description>Reporter for the metrics library which reports into Elasticsearch</description>
```
Comment (chibenwa) [inline third-party/nerdynick-elasticsearch-metrics-reporter/pom.xml]: Can we add in the description the original repo URL (attribution) ?

**Comment by rouazana** [comment]
Most comments are related to code I just imported. I did that because I just considered that as an external lib. And we don't review all the code of all the external libs we are using (shoud we? ^^).
This library is imported just for the purpose of dependency management. Being not maintained anymore, we need to get it in our source code to be able to use it. Alternatives are:
1. just compile the jar and put it with James
2. just import this library and don't care (what I did)
3. fork the original repo and add some features to it (ES 6 support, update of libraries, etc.)

I'm really uncomfortable making a local fork (I mean just inside the main James repository) and fix things inside it. We loose all the commit history, and (back/front)porting things will be a nightmare.

If you think we need to maintain a real version of this lib, and put it under org.james namespace to be able to release it, maybe we need to really fork the project? Then I'm ok considering your remarks.

File: third-party/nerdynick-elasticsearch-metrics-reporter/pom.xml (hunk 5 above)
Comment (rouazana) [reply third-party/nerdynick-elasticsearch-metrics-reporter/pom.xml]: it's only a test dependency, does it harm?

Comment (mbaechler) [reply third-party/nerdynick-elasticsearch-metrics-reporter/pom.xml]: yes it does, it makes it harder to enable logging on debug.
Having logback even in test scope is a good strategy

**Comment by rouazana** [comment]
see #2002

**Pull request closed without merging**